File: 0518-coin-change-ii/0518-coin-change-ii.py
class Solution:
    def change(self, amount: int, coins: List[int]) -> int:
        dic = defaultdict(int)
        for i in range(len(coins)):dic[(0,i)] = 1
        
        for amt in range(1,amount+1):
            for i in range(len(coins)-1,-1,-1):
                if amt - coins[i] >= 0:
                    dic[(amt,i)] += (dic[(amt-coins[i],i)] + dic[(amt,i+1)])
                else: dic[(amt,i)] += dic[(amt,i+1)]
        print(dic[2,1])
        return dic[(amount,0)]
            
                    
                

File: 0518-coin-change-ii/0518-coin-change-ii.cpp
class Solution {
    int dp[5001];
public:
    int change(int amount, vector<int>& coins) {
        memset(dp,0,sizeof (amount+1)*4);
        dp[0] = 1;
        for(int i = 0; i < coins.size();i++)
            for(int j = coins[i]; j <= amount;j++)dp[j] += dp[j - coins[i]];
        return dp[amount];
        
    }
};